### ISSUE TITLE
Clarify the legal status

### ISSUE THREAD

**Issue by bugaevc** [skypeopensource/skypeopensource2#2]
The README should clearly state:
*  whether or not this client is a complete **re-implementation**
* whether or not it includes **disassembled** code from the other Skype clients
* whether or not it is a **clean-room reverse-engineered** code, like Wine/ReactOS
* why the specifies **license is valid** and legal to use for the project
* what are the **legal complications**, if any, to developing the project and to using it personally or commercially

**Comment by tmikaeld** [comment]
Considering it's secrecy, I'd bet it's not legal

**Comment by lgommans** [comment]
This repository is called `skypeopensource2`, made me wonder if where was a `1`. Turns out it has been taken down via DMCA via this notice: https://github.com/github/dmca/blob/master/2012/2012-05-01-microsoft.markdown

Is this a copy of the original repository? Or a new method of implementation that does comply with the law?

**Comment by skypeopensource** [comment]
@lgommans No, its new repo and code. Original "skypeopensource" was banned because i use it to share patched skype binary needed for reversing purposes. Which is no-no action in any case.

**Comment by skypeopensource** [comment]
@bugaevc 

> whether or not this client is a complete re-implementation

Yes. Its complete re-implementation of proprietary Skype Network protocol.

> whether or not it includes disassembled code from the other Skype clients

Well, may be a little :) unpack-4142.c and pack-4142.c is HexRays decompiled code. And should be rewrited... But i dont have time for this now/

> whether or not it is a clean-room reverse-engineered code, like Wine/ReactOS

No. Not clean room, not chinese wall.

> why the specified license is valid and legal to use for the project

Its my own code, i write it, i am the author of most code. And hold all copyrights rights.

> what are the legal complications, if any, to developing the project and to using it personally or commercially

1) The pack-4142.c/unpack41-42.c mentioned above. 
2) Also, it contain MIRACL for RSA and big numbers multiply arithmetics. 

> MIRACL Cryptographic SDK: Multiprecision Integer and Rational Arithmetic Cryptographic Library is a C software library that is widely regarded by developers as the gold standard open source SDK for elliptic curve cryptography (ECC). http://miracl.com
https://github.com/miracl/MIRACL

For commercial usage and license bring me email to: skypeopensource@gmail.com

**Comment by bugaevc** [comment]
@skypeopensource you should add that to the README -- and it should be clear and unambiguous

> Well, may be a little :) unpack-4142.c and pack-4142.c is HexRays decompiled code. And should be rewrited... But i dont have time for this now/

If so, how can you claim that it's a complete LGPL-ed re-implementation?

> Its my own code, i write it, i am the author of most code. And hold all copyrights rights.

So, if it includes decompiled stuff, how can it be? I bet the Skype license prohibits decompiling it at all, let alone publishing results under LGPL.

**Comment by skypeopensource** [comment]
@bugaevc 

> If so, how can you claim that it's a complete LGPL-ed re-implementation?

Well, not complete. Or complete except this files. You are right. I can remove it.

**Comment by skypeopensource** [comment]
@bugaevc 

> So, if it includes decompiled stuff, how can it be? I bet the Skype license prohibits decompiling it at all, let alone publishing results under LGPL.

Its contain very little of this stuff. In compare to other project code. Imagine, ts just like zip/unzip code. Other files is created by reversing and analyzing network protocol flow.

**Comment by bugaevc** [comment]
@skypeopensource 

> Its contain very little of this stuff. In compare to other project code. Imagine, ts just like zip/unzip code. Other files is created by reversing and analyzing network protocol flow.

Look, I _do_ value your work. It (the free Skype client) _may become_ one of the best things that's happening to Linux and it _may_ drastically increase Linux desktop adoption.

But it should be perfectly clean (copyright and license-wise) and legal. Free -- as in freedom -- software is never about stealing stuff. Free software shouldn't be "mostly legal".

**Comment by marcan** [comment]
FWIW, isn't this hopelessly obsolete anyway? As far as I know, the old Skype protocol is effectively dead; newer versions of Skype use a reincarnation of the MSN Messenger protocol. There are some docs on that [here](https://github.com/msndevs/protocol-docs/wiki).

**Comment by ultramancool** [comment]
Legal status will really depend on your country. Reverse engineering for compatibility, even if done without a Chinese wall or similar tactic is still perfectly legal in many places. This is an extremely good effort and would probably need to be the first step even if a Chinese-wall style solution were to be developed from it later. So thanks for your hard work.

**Comment by riverar** [comment]
Careful, with the VEST copyright bits in place, this project has no legitimate use or license.

**Comment by marcan** [comment]
Also, pack-4142.c/unpack41-42.c aren't the only decompiled code. util.c, crc32_util.c, dh_384.c, decode41.c all contain obviously decompiled code, and that's just judging by the variable names. Which makes me wonder about the rest of the code. Decompiling code and "massaging" it until it no longer looks like decompiled code isn't legal. You can reverse engineer without a clean-room approach and be in the clear in many jurisdictions, but it takes mental process similar to cleanrooming where you build a spec, either explicitly or mentally, and code against that. Starting with decompiled code and cleaning it up is not that process. That's a derivative work, not a "reimplementation".

This whole thing is a commendable reverse engineering effort, but not distributable in its current form, certainly not licensable as LGPLv3 as a whole (not only the decompiled bits, but also MIRACL seems to be AGPLv3 which would make the whole project AGPLv3 just by itself).

**Comment by marcan** [comment]
Addendum: it seems VEST was fined 5000 euro for the code that's in this repo. I wouldn't go anywhere near it. http://www.zdnet.fr/actualites/decompiler-skype-5000-euros-d-amende-et-6-mois-de-sursis-39817536.htm

**Comment by bugaevc** [comment]
@marcan thanks for your overview! Am I getting it right -- the new protocol (MSNP24) is kinda-known and there are open-source implementations? If so, why there's no [popular] third-party Skype client?

**Comment by skypeopensource** [comment]
Specially for https://geektimes.ru/post/280068/

This comment https://github.com/skypeopensource/skypeopensource2/issues/2#issuecomment-244595285

Written by me, and it is reply to Issue Ticket opened by @bugaevc to tell my about some difficulties about copyrights and other stuff.

Only me, Efim Bushmanov is author of biggest part of this code base. Except some three files mentioned in README.md file of this repo.

**Comment by bugaevc** [comment]
I am in no way related to this project other than participating in a couple of discussions around here and filing the two issues.

The comment in question was indeed written by @skypeopensource in response to me.

@alizar, please learn to check your sources. I demand you remove any mention of me from the article ASAP.

**Comment by marcan** [comment]
To elaborate on just how much of the code in here is not freely distributable:

* [dh_384.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/skycontact3/skycontact3/dh_384.c#L19) - this code is clearly Hex-Rays decompiler output.
* [util.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/skycontact3/skycontact3/util.c#L203) - this code contains x86 register names and is obviously partially but not completely cleaned up output from decompilation.
* [crc32_util.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/crc32_util.c#L128) - same, x86 register names and also the first table follows the IDA Pro disassembler naming scheme and has the original memory address in the name.
* [decode41.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/skyrelay3/skyrelay3/decode41.c#L2434) - blatant Hex-Rays output here again, the function name even has the address. This entire file is decompiler output in various stages of cleanup - just search for "eax", there are a gazillion hits, but the linked function isn't even cleaned up at all, it's raw decompiler output. This is a 3700-line file that is clearly not redistributable or licensable under the LGPL.

And some more subtle ones:

* [ext42.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42.c#L672) - "stack mess"? Obviously garbage variables left over from decompilation. The function names are also the kind of thing a reverse engineer would write during analysis, not something a programmer would write while writing new code. And I know Hex-Rays idioms when I see them, and even though you've managed to get rid of all the blatant x86 register names, this is clearly derived from Hex-Rays output.
* [ext42_2.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42_2.c) - more of the same
* [ext42_checkblob.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42_checkblob.c) - more of the same
* [ext42_getobj00.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42_getobj00.c) - more of the same
* [ext42_getobj00seq.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42_getobj00seq.c) - more of the same
* [ext42_getobj00ip.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/ext42_getobj02ip.c) - more of the same

And then there are the big constants and tables, which could also be a liability, depending on what they are and whether they are strictly required for interoperability.

That's for decompiler output. Now, as for VEST copyrights:

* [skype_rc4.h](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay3/goodrecvrelay3/skype/skype_rc4.h)
* [skype_login.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/skyauth3/skyauth3/skype_login.c)
* [skype_basics.h](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/skype_basics.h)
* [unpack-4142.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay3/goodrecvrelay3/skype/unpack-4142.c) - both VEST copyright *and* decompiler output, i.e. this file has contradictory copyright information, since it can't both legitimately be copyright VEST and decompiled output, i.e. a derivative work of Skype.  1k lines of code.
* [pack-4142.c](https://github.com/skypeopensource/skypeopensource2/blob/4870457f725b332ec52d6d72e688c9beddbdd87b/goodrecvrelay4_dll/goodrecvrelay4_dll/skype/pack-4142.c) - same as above. 1k lines of code.

There are no "excepts" or "biggest part" in open source. Either the code is legitimately licensable under the LGPL, or it isn't. This isn't even remotely close, nor will it ever be, with this history. In fact, by starting here, you've pretty much irreversibly tainted the code as it is evidently that it is a derivative work of the Skype binary.

So, as I said, this is very commendable reverse engineering of the Skype protocol, but it is *not* free software, it is *not* LGPL, it is *not* redistributable, and it is *not* something that should be on GitHub, and, if Microsoft cares, they will have a perfectly legitimate claim to filing a DMCA request and having this repo taken down.

**Comment by bugaevc** [comment]
@skypeopensource please contact @alizar via Geektimes or in some other way and make sure he understands you are the author of this project, not me. Thanks.

(I'm not sure the username on GitHub belongs to him.)

**Comment by ValdikSS** [comment]
@bugaevc Just contacted him.

**Comment by ValdikSS** [comment]
@bugaevc the article is fixed, now it doesn't mention you.

**Comment by ValdikSS** [comment]
@bugaevc Since the problem is resolved I'm going to delete my comment as it's off topic.

**Comment by bugaevc** [comment]
@ValdikSS thanks a lot. My name is still in the tags (I know, "nobody reads the tags" :smile: ), and in the last paragraph the verbs are in plural, but I think I can live with that.

**Comment by ValdikSS** [comment]
@bugaevc 
>My name is still in the tags 

Aaand it's gone!

**Comment by bugaevc** [comment]
@ValdikSS great, thanks again!

**Comment by alizar** [comment]
@bugaevc Everything is fixed. Sorry for inconvinience...

**Comment by skypeopensource** [comment]
https://github.com/skypeopensource/skypeopensource2/issues/2#issuecomment-244617932
@marcan 

As long as i remember, it not have purpose of reversing for interoperability. So, they got a fine.

**Comment by skypeopensource** [comment]
https://github.com/skypeopensource/skypeopensource2/issues/2#issuecomment-244626692
@marcan 

Thank you.

**Comment by 4577** [comment]
>Addendum: it seems VEST was fined 5000 euro for the code that's in this repo. I wouldn't go anywhere near it. http://www.zdnet.fr/actualites/decompiler-skype-5000-euros-d-amende-et-6-mois-de-sursis-39817536.htm

This case was a bit complex, court first discharged them because it considered that the released code was akin to a math formula and couldn't be copyrighted, but then sentenced them in appeal for counterfeit primarily because he intended to make larger than interoperability stuff out of it. Anyway the VEST guy is serving a sentence for sexual enslavement so I don't know how well he handled the trial, and you should keep in mind that this was not American nor Russian but European law.

The "Flux Capacitor" (RC4 key seeder) logic isn't that complex, it's just a large chunk of branches with some visible patterns so you could likely reimplement it off publicly available info while plausibly not reproducing too much of its specific essence.
